variables/dependencies.go

```diff
@@ -0,0 +1,129 @@
+type Dependency struct {
+	Name                  string
+	TemplateFolder        string
+	OutputFolder          string
+	DontInheritVariables  bool
+	Variables             []Variable
+}
+
+// Get all the variables in this dependency, namespacing each variable with the name of this dependency
+func (dependency Dependency) GetNamespacedVariables() []Variable {
+	variables := []Variable{}
+
+	for _, variable := range dependency.Variables {
+		variableNameForDependency := fmt.Sprintf("%s.%s", dependency.Name, variable.Name())
+		variables = append(variables, variable.WithName(variableNameForDependency))
+	}
+
+	return variables
+}
+
+// Given a unique variable name, return a tuple that contains the dependency name (if any) and the variable name.
+// Variable and dependency names are split by a dot, so for "foo.bar", this will return ("foo", "bar"). For just "foo",
+// it will return ("", "foo").
+func SplitIntoDependencyNameAndVariableName(uniqueVariableName string) (string, string) {
+	parts := strings.SplitAfterN(uniqueVariableName, ".", 2)
+	if len(parts) == 2 {
+		// The split method leaves the character you split on at the end of the string, so we have to trim it
+		return strings.TrimSuffix(parts[0], "."), parts[1]
+	} else {
+		return "", parts[0]
+	}
+}
+
+// Given a map of key:value pairs read from a Boilerplate YAML config file of the format:
+//
+// dependencies:
+//   - name: <NAME>
+//     template-folder: <TEMPLATE_FOLDER>
+//     output-folder: <OUTPUT_FOLDER>
+//
+//   - name: <NAME>
+//     template-folder: <TEMPLATE_FOLDER>
+//     output-folder: <OUTPUT_FOLDER>
+//
+// This method takes the data above and unmarshals it into a list of Dependency objects
+func UnmarshalDependenciesFromBoilerplateConfigYaml(fields map[string]interface{}) ([]Dependency, error) {
+	unmarshalledDependencies := []Dependency{}
+	dependencyNames := []string{}
+
+	listOfFields, err := unmarshalListOfFields(fields, "dependencies")
+	if err != nil {
+		return unmarshalledDependencies, err
+	}
+
+	for _, fields := range listOfFields {
+		dependency, err := UnmarshalDependencyFromBoilerplateConfigYaml(fields)
+		if err != nil {
+			return unmarshalledDependencies, err
+		}
+
+		if util.ListContains(dependency.Name, dependencyNames) {
+			return unmarshalledDependencies, errors.WithStackTrace(DuplicateDependencyName(dependency.Name))
+		}
+		dependencyNames = append(dependencyNames, dependency.Name)
+
+		unmarshalledDependencies = append(unmarshalledDependencies, *dependency)
+	}
+
+	return unmarshalledDependencies, nil
+}
+
+// Given a map of key:value pairs read from a Boilerplate YAML config file of the format:
+//
+// name: <NAME>
+// template-folder: <TEMPLATE_FOLDER>
+// output-folder: <OUTPUT_FOLDER>
+//
+// This method takes the data above and unmarshals it into a Dependency object
+func UnmarshalDependencyFromBoilerplateConfigYaml(fields map[string]interface{}) (*Dependency, error) {
```

Comment:
I actually haven't worked with YAML and Scala, but I have worked with JSON and Scala, and since those are similar, it means I'd have a bunch of advantages:

* Go doesn't allow you to set default values in structs. Therefore, after unmarshalling in Go, I'd end up with a bunch of blank fields, and would have to manually add validation logic all over the place to check for that and fill in defaults. In Scala, you can easily specify a default value.
* With reflection-based JSON libraries (e.g. [Jackson](https://github.com/FasterXML/jackson)), you can to specify a custom deserializer for a single type ([example](http://stackoverflow.com/a/19167145/483528)) or a single field within a type. In the deserializer, you implement a single method that returns the deserialized value. Simple, intuitive. With Go, you you can specify a custom deserializer for an entire type, but not individual fields. And when you implement it, the [API is very unintuitive](https://github.com/go-yaml/yaml/blob/v2/yaml.go#L32) and requires you to mess with pointers.
* With more manual JSON libraries (e.g. [Play JSON](https://www.playframework.com/documentation/2.5.x/ScalaJsonHttp)), you have strong tools for expressing how to parse each field in the JSON into the exact type you want. With go-yaml's clunky API, you don't.
* With macro-based JSON libraries (e.g. [Play JSON, again](https://www.playframework.com/documentation/2.5.x/ScalaJsonAutomated)), you can have a lot of the hard work done by a macro, and only do the custom stuff manually.
* Finally, due to Go's lack of support for generics and convenient methods for converting types, doing this sort of unmarshalling work involves a lot of convert `map[interface{}]interface{}` to `map[string]string` style work by hand.